core/WFCORE-3435_capability_requirements_deployments.adoc

```diff
@@ -0,0 +1,184 @@
+Deployments that are activating services must be able to work with capabilities too.
+
+== Issue Metadata
+
+* WildFly issue: https://issues.jboss.org/browse/WFCORE-3435
+* Affected projects or components:
+** https://github.com/wildfly/wildfly-core[Wildfly Core]
+** https://github.com/jboss-msc/jboss-msc[JBoss Modular Service Container]
+** https://github.com/wildfly-swarm/wildfly-swarm[WildFly Swarm]
+
+== Requirements
+
+* A deployment should be able to activate a service that requires a capability.
+* The mechanism to let user interact with capabilities should be similar to the one used for services (`ServiceActivator` implementation loaded by Java `ServiceLoader`).
+
+== Non-Requirements
+
+* Current use of `ServiceActivator` is valid and must not be affected by this change.
+* `ServiceActivator` API remains valid and will not be deprecated.
+
+== WildFly Core Requirements
+
+WildFly Core is providing the `CapabilityServiceSupport` to support capability integration in service implementations.
+Capability is a concept from WildFly Core, while `ServiceActivator` is an API provided by MSC. A new API that allows deployments to activate capabilities must be provided by WildFly Core.
+
+* The `CapabilityServiceSupport` API should be exposed to deployments that are activating services so they can integrate with capabilities too.
+* User deployments must be processed and any services that require capabilities be activated.
+
+== Current design
+
+Currently, an user must implement `org.jboss.msc.service.ServiceActivator` and install MSC Services using the `ServiceActivatorContext` parameter.
+Its implementation is loaded by the Java ServiceLoader mechanism and its services are activated during deployment.
+
+[source]
+----
++-----------------------------------+       +--------------------+
+|             << MSC >>             |       |     << User >>     |
+|         ServiceActivator          |<|-----| MyServiceActivator |
++-----------------------------------+       +--------------------+
+| activate(ServiceActivatorContext) |
++-----------------------------------+
+        ^
+       / \
+        |
++---------------------------------------+
+|             << MSC >>                 |
+|       ServiceActivatorContext         |
++---------------------------------------+
+| getServiceTarget(): ServiceTarget     |
+| getServiceRegistry(): ServiceRegistry |
++---------------------------------------+
+----
+
+== Proposed Design
+
+We would need to introduce a new API so that an user could activate services that depends on capabilities.
+This API is provided by WildFly Core (in its wildfly-controller artifact).
+
+The user must implement the `org.jboss.as.controller.CapabilityActivator` interface.
+This interface provides a `org.jboss.as.controller.CapabilityActivatorContext` interface that extends `org.jboss.msc.service.ServiceActivatorContext` and add a method to access the `org.jboss.as.controller.capability.CapabilityServiceSupport`.
```

Comment:
Why do we need all these new types? Could we not simply add the `hasCapability`/`getCapability` methods to `ServiceActivatorContext` directly?

Reply:
`CapabilityServiceSupport` is an existing class that DUP can already uses.

I don't like the idea to introduce in MSC API the notion of "capabilities" that are only relevant to WildFly Core.
If you think it is ok to add to the `ServiceActivatorContext` all the methods from `CapabilityServiceSupport` that will ease my work but it does not feel conceptually clean.

For reference, all the methods are:
```
boolean hasCapability(String capabilityName);
<T> T getCapabilityRuntimeAPI(String capabilityName, Class<T> apiType) throws NoSuchCapabilityException;
<T> T getCapabilityRuntimeAPI(String capabilityBaseName, String dynamicPart, Class<T> apiType) throws NoSuchCapabilityException;
ServiceName getCapabilityServiceName(String capabilityName);
ServiceName getCapabilityServiceName(String capabilityBaseName, String ... dynamicParts);
```

Reply:
OK, I forgot that that `ServiceActivatorContext` was in MSC.